Convert cts-fencing to use XML output

clumens opened the pull request.

kgaillot commented:
Looks good, but I don't want to make xmllint a hard dependency -- skip the check if it's not available. There are alternatives to xmllint as well, so even better, make it easy to drop in those later and use whatever's available.

kgaillot commented:
With python 3:

Traceback (most recent call last):
  File "/usr/share/pacemaker/tests/cts-fencing", line 1509, in <module>
    main(sys.argv)
  File "/usr/share/pacemaker/tests/cts-fencing", line 1501, in main
    tests.run_tests()
  File "/usr/share/pacemaker/tests/cts-fencing", line 559, in run_tests
    test.run()
  File "/usr/share/pacemaker/tests/cts-fencing", line 463, in run
    self.run_cmd(cmd)
  File "/usr/share/pacemaker/tests/cts-fencing", line 391, in run_cmd
    xmllint_out = xmllint.communicate(input=output)
  File "/usr/lib64/python3.6/subprocess.py", line 863, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib64/python3.6/subprocess.py", line 1519, in _communicate
    input_view = memoryview(self._input)
TypeError: memoryview: a bytes-like object is required, not 'str'

clumens commented:
It's been a while since I've dealt with python2 vs. python3 strings and bytes, but here is one possibility.  This works for both versions:
```diff --git a/cts/cts-fencing.in b/cts/cts-fencing.in
index 0f9bd6ba3..78fe91067 100644
--- a/cts/cts-fencing.in
+++ b/cts/cts-fencing.in
@@ -115,11 +115,19 @@ def find_validator(rng_file):
         return None
 
 
-def pipe_output(pipes, stdout=True, stderr=False):
+def pipe_communicate(pipes, stdout=True, stderr=False, stdin=None):
     """ Wrapper to get text output from pipes regardless of Python version """
 
     output = ""
-    pipe_outputs = pipes.communicate()
+
+    if stdin:
+        if sys.version_info < (3,):
+            pipe_outputs = pipes.communicate(input=stdin)
+        else:
+            pipe_outputs = pipes.communicate(input=stdin.encode())
+    else:
+        pipe_outputs = pipes.communicate()
+
     if sys.version_info < (3,):
         if stdout:
             output = output + pipe_outputs[0]
@@ -138,7 +146,7 @@ def output_from_command(command):
 
     test = subprocess.Popen(shlex.split(command), stdout=subprocess.PIPE)
     test.wait()
-    return pipe_output(test).split("\n")
+    return pipe_communicate(test).split("\n")
 
 
 def localname():
@@ -361,7 +369,7 @@ class Test(object):
         else:
             return CrmExit.OK
 
-        output = pipe_output(test, stderr=True)
+        output = pipe_communicate(test, stderr=True)
         if self.verbose:
             print(output)
 
@@ -388,13 +396,13 @@ class Test(object):
                 print("\nRunning: "+" ".join(cmd))
 
             xmllint = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-            xmllint_out = xmllint.communicate(input=output)
+            output = pipe_communicate(xmllint, stderr=True, stdin=output)
 
             if self.verbose:
-                print(xmllint_out[1])
+                print(output)
 
             if xmllint.returncode != 0:
-                raise XmlLintError(xmllint_out[1])
+                raise XmlLintError(output)
 
 
     def count_negative_matches(self, outline):
```

kgaillot commented:
Works for me

clumens commented:
One other question - most of the tests look at log files instead of the XML on stdout to figure out if they should pass or not.  Would it be worth going through them all and converting them to look at the XML where possible?

kgaillot commented:
No, the log patterns are useful for verifying internal pacemaker-fenced operation that's invisible to the client.

kgaillot commented on cts/cts-fencing.in:
```diff
@@ -188,6 +203,16 @@ class OutputFoundError(TestError):
 
 
+class XmlLintError(TestError):
+    """ Exception raised when xmllint fails """
```
Since we're leaving open the possibility of other validators in the future, let's make this generic e.g. XmlValidationError

kgaillot commented on cts/cts-fencing.in:
```diff
@@ -356,6 +382,28 @@ class Test(object):
+            if os.path.exists("%s/cts-fencing.in" % TEST_DIR):
+                rng_file = "xml/api/api-result.rng"
+            else:
+                rng_file = "/usr/share/pacemaker/api-result.rng"
```
The schema directory is configurable at both build-time and run-time, so we can't hard-code it. Use os.environ['PCMK_schema_directory'] if it exists (whether running from checkout or not), and the checkout location (as you have it above) or @CRM_SCHEMA_DIRECTORY@ otherwise (I like to define constants for such build options at the top so all the funky substitution syntax is in one place).

kgaillot commented on cts/cts-fencing.in:
```diff
@@ -356,6 +382,28 @@ class Test(object):
+            if self.verbose:
+                print("\nRunning: "+" ".join(cmd))
+
+            xmllint = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
use generic variable name here too

kgaillot commented on cts/cts-fencing.in:
```diff
@@ -108,11 +109,35 @@ def update_path():
+
+
+def rng_directory():
+    if os.environ.has_key("PCMK_schema_directory"):
```
python 3 doesn't have has_key() :)

you can use either "PCMK_schema_directory" in os.environ or try: return ... and handle the exception

clumens replied:
Of course it doesn't.

However, I really should be using the `in` syntax here, which appears to work for both python2 and python3.  So:
```
if "PCMK_schema_directory" in os.environ:
...
```

It's good to see that I'm forgetting all this basic python stuff after being out of it for just a few short months.
